---
title: "GWAS Pipeline Report"
---

```{r include=FALSE}
#!/usr/bin/env Rscript
args<-commandArgs(TRUE)
outdir <- args[1]
```

```{r include=FALSE}
suppressPackageStartupMessages(library(SeqArray))
suppressPackageStartupMessages(library(knitr))
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(EBImage))
```

##QC
The number of SNPs included in the gds file for each chromosome and total:
```{r include=FALSE}
snp.n.gds <- c()
for(i in 1:22){
  gds <- seqOpen(paste0(outdir,"/GDS/gds_files/chr_",i,".gds"))
  snp.n.gds <- c(snp.n.gds, length(seqGetData(gds, "variant.id")))
  seqClose(gds)
}
snp.n.gds <- c(snp.n.gds, sum(snp.n.gds))
chr <- c(paste0("chr", 1:22), "Total")
dat.gds <- data.frame(count=snp.n.gds)
rownames(dat.gds) <- chr
```
```{r}
kable(t(dat.gds))
```

The number of SNPs analyzed in the regression model:
```{r include=FALSE}
snp.n.r <- c()
for(i in 1:22){
  try(result <- fread(paste0(outdir,"/Association_Test/gwas/chr_",i,".csv"), stringsAsFactors = F, header = T))
  try(result <- fread(paste0(outdir,"/Association_Test/gwla/chr_",i,".txt"), sep="\t", stringsAsFactors = F, header = T))
  snp.n.r <- c(snp.n.r, dim(result)[1])
}
snp.n.r <- c(snp.n.r, sum(snp.n.r))
missing.rate <- round(1-snp.n.r/snp.n.gds, digits = 3)
dat.r <- data.frame(count=snp.n.r, missing.rate=ifelse(missing.rate < 0, 0, missing.rate))
rownames(dat.r) <- chr
```
```{r}
kable(t(dat.r))
```

The number of SNPs with valid analysis results (2NxMAF>1):
```{r include=FALSE}
snp.n.f <- c()
for(i in 1:22){
  result <- fread(paste0(outdir,"/Summary_Plot/merge_by_chr/chr_",i,"_caf_annotated.csv"))
  snp.n.f <- c(snp.n.f, dim(result)[1])
}
snp.n.f <- c(snp.n.f, sum(snp.n.f))
dat.f <- data.frame(count=snp.n.f, missing.rate=round(1-snp.n.f/snp.n.r, digits = 3))
rownames(dat.f) <- chr
```
```{r}
kable(t(dat.f))
```


##PCA plot
```{r include=FALSE}
try(pc1 <- readImage(paste0(outdir, "/PCA_GRM/pcair/PC1vsPC2.png")))
try(pc2 <- readImage(paste0(outdir, "/PCA_GRM/pcair/PC3vsPC4.png")))
```
```{r}
par(mfrow=c(1,2))
try(display(pc1))
try(display(pc2))
```

##Kinship plot
```{r include=FALSE}
try(kinship <- readImage(paste0(outdir, "/PCA_GRM/pcrelate/kinship.png")))
```
```{r}
try(display(kinship))
```

##QQ and Manhattan plot for all SNPs passed filter
```{r include=FALSE}
try(qq1 <- readImage(paste0(outdir, "/Summary_Plot/qq_manhattan/qqplot.png")))
```
```{r}
try(display(qq1))
```
```{r include=FALSE}
try(manhattan1 <- readImage(paste0(outdir, "/Summary_Plot/qq_manhattan/qqplot_MAF_0.1.png")))
```
```{r}
try(display(manhattan1))
```

##QQ and Manhattan plot for SNPs passed filter and MAF threshold
```{r include=FALSE}
try(qq2 <- readImage(paste0(outdir, "/Summary_Plot/qq_manhattan/manhattan_plot.png")))
```
```{r}
try(display(qq2))
```
```{r include=FALSE}
try(manhattan2 <- readImage(paste0(outdir,"/Summary_Plot/qq_manhattan/manhattan_plot_MAF_0.1.png")))
```
```{r}
try(display(manhattan2))
```

##Annotated top results
```{r include=FALSE}
annotated.results <- fread(paste0(outdir, "/Annotation/annotated_results/top_snps_annotation.csv"), header=T, stringsAsFactors=F)
```
```{r}
kable(annotated.results)
```
